# Update Null Values

In [17]:
import pandas as pd
import numpy as np

## Load Data

In [18]:
data = pd.read_csv('../final_project_mod.csv')

In [19]:
data.head()

x0        x1         x2        x3        x4         x5         x6  \
0 -0.166563 -3.961588   4.621113  2.481908 -1.800135   0.804684   6.718751   
1 -0.149894 -0.585676  27.839856  4.152333  6.426802  -2.426943  40.477058   
2 -0.321707 -1.429819  12.251561  6.586874 -5.304647 -11.311090  17.812850   
3 -0.245594  5.076677 -24.149632  3.637307  6.505811   2.290224 -35.111751   
4 -0.273366  0.306326 -11.352593  1.676758  2.928441  -0.616824 -16.505817   

          x7        x8        x9  ...        x41       x42       x43  \
0 -14.789997 -1.040673 -4.204950  ...  -1.497117  5.414063 -2.325655   
1  -6.725709  0.896421  0.330165  ...  36.292790  4.490915  0.762561   
2  11.060572  5.325880 -2.632984  ...  -0.368491  9.088864 -0.689886   
3 -18.913592 -0.337041 -5.568076  ...  15.691546 -7.467775  2.940789   
4  27.532281  1.199715 -4.309105  ... -13.911297 -5.229937  1.783928   

        x44       x45        x46       x47       x48        x49  y  
0  1.674827 -0.264332  60.781427 -7.689696  0.151589  -8.040166  0  
1  6.526662  1.007927  15.805696 -4.896678 -0.320283  16.719974  0  
2 -2.731118  0.754200  30.856417 -7.428573 -2.090804  -7.869421  0  
3 -6.424112  0.419776 -72.424569  5.361375  1.806070  -7.670847  0  
4  3.957801 -0.096988 -14.085435 -0.208351 -0.894942  15.724742  1  

[5 rows x 51 columns]

## Locate Null Values: Column by Row - Populate Dict

Column by row where value is null

In [3]:
data_notnull = data.notnull()
nullValues = {}
for column in data:
    nullValues.update({column: data_notnull[data_notnull[column]==False].index.to_list()})

Display the ten index's (Rows) where there is missing values (NaN) in attribute x0.

In [4]:
nullValues['x0'][0:10]

[1282, 2290, 7491, 15484, 22497, 23192, 28954, 35999, 53729, 72318]

Proving there are NaN's within the first ten locations within the x0 column

In [5]:
# data.loc[ROW, COLUMN]
data.loc[nullValues['x0'][0:10] ,'x0']

1282    NaN
2290    NaN
7491    NaN
15484   NaN
22497   NaN
23192   NaN
28954   NaN
35999   NaN
53729   NaN
72318   NaN
Name: x0, dtype: float64

## Distinct Rows With Missing Value

In [6]:
unique_index = []
for key, value in nullValues.items():
    for i in value:
        if i not in unique_index: 
            unique_index.append(i)
unique_index.sort()

In [7]:
print('There are', len(unique_index), 'DISTINCT rows with a NaN in at least one of the attributes.')

There are 1608 DISTINCT rows with a NaN in at least one of the attributes.


## Replace NaN With arithmetic averages (Mean, Median, Mode, etc)

In [14]:
for column in data.columns:
    print(column, data[column].mean())

Unnamed: 0 79999.5
x0 -0.0010279116076364367
x1 0.0013583132941497986
x2 -1.150144637185299
x3 -0.02463729930505851
x4 -0.0005489999661714708
x5 0.01358155232776665
x6 -1.6706701553089665
x7 -7.692794926068938
x8 -0.030540285688330632
x9 0.005462093779336137
x10 0.002253454768289208
x11 0.030232406883283356
x12 -1.334402055040244
x13 0.007669015129559942
x14 0.008103587049699257
x15 0.0012147158213739108
x16 0.006223144342518333
x17 0.012039895626624077
x18 0.01269395995231238
x19 0.024555427951995574
x20 0.2990740609444575
x21 -0.02913741802162567
x22 0.008400020260830079
x23 0.7220283102008699


TypeError: can only concatenate str (not "int") to str

What is the mode for column x0

In [12]:
data['x49'].mean()

-0.6742238327007368

Return to nan within the first location (row 1282) within the x0 column to see that it has been replace with the mode

In [12]:
data_copy.loc[nullValues['x0'][0] ,'x0']

-1.5926345910622286